File: Dynamic_Programming/1959. Minimum Total Space Wasted With K Resizing Operations.cpp
/*
1959. Minimum Total Space Wasted With K Resizing Operations
*/

// 2025.03.22 優化遞推 TC:O(n * k * n) SC:O(n * k)
class Solution {
public:
    int minSpaceWastedKResizing(vector<int>& nums, int k) {
        int n = nums.size();
        int f[n+1][k+1];
        for (int i = 0; i <= n; i++)
            fill(f[i], f[i] + k + 1, INT_MAX/2);
        f[0][0] = 0;
        int mx = 0, tot = 0;
        for (int i = 0; i < n; i++) {
            mx = max(mx, nums[i]);
            tot += nums[i];
            f[i + 1][0] = mx * (i + 1) - tot;
        }
        
        for (int i = 0; i < n; i++) {
            int groups = min(i + 1, k);
            // j 表示已完成的切分數（0-indexed，j+1 段）
            for (int j = 0; j < groups; j++) {
                mx = 0, tot = 0;
                for (int t = i; t >= 0; t--) {
                    // 如果 t < j，表示前 t 個元素不足以分成 j+1 段，提前結束
                    if (t + 1 < j) break;
                    mx = max(mx, nums[t]);
                    tot += nums[t];
                    f[i + 1][j + 1] = min(f[i + 1][j + 1], f[t][j] + mx * (i - t + 1) - tot);
                }
            }
        }
        return f[n][k];
    }
};



// TC:O(n*k*n) SC:O(n*k)
class Solution {
public:
    int minSpaceWastedKResizing(vector<int>& nums, int k) {
        int n = nums.size();
        int dp[n][k+1];
        for(int i = 0; i < n; i++){
            for(int j = 0; j <= k; j++){
                dp[i][j] = INT_MAX/2;
            }
        }
        //初值設定 都沒做操作的代價
        int mx = 0;
        int sum = 0;
        for(int i = 0; i < n; i++){
            mx = max(mx, nums[i]);
            sum += nums[i];
            dp[i][0] = mx*(i+1) - sum;
        }

        for(int i = 1; i < n; i++){
            for(int j = 1; j <= min(i,k); j++){
                int mx = 0;
                int intervalSum = 0;
                //反著走才不用每次都重新走一遍
                for(int s = i; s >= 1; s--){
                    mx = max(mx, nums[s]);
                    intervalSum += nums[s];
                    dp[i][j] = min(dp[i][j], dp[s-1][j-1] + mx*(i-s+1) - intervalSum);
                }
            }
        }

        int res = INT_MAX/2;
        for(int j = 0; j <= k; j++){
            res = min(res, dp[n-1][j]);
        }

        return res;
    }
};

/*
此題與1043 1105相似
此題問說如果你可以走一步就改變一個大小讓他去大於等於某段目前或者之後的區間
問說改動次數最多k次  改完後跟每次改動後他所負責的那段區間元素的差最小值為多少?
dp[i][j] : for the first i element, we used j operations. the minimum total space wasted 
會發現每次你要改動時都會有改動他可以cover的區間 那在他之前的最小值加上改動後的space wasted又可以構成新答案  -> dp
dp[i][j] = dp[s-1][j-1] + nums[s]*(i-s+1) - intervalSum;
                           wasteSpace
*/

File: Dynamic_Programming/1959. Minimum Total Space Wasted With K Resizing Operations.py
# TC:O(n * k * n) SC:O(n * k)
class Solution:
    def minSpaceWastedKResizing(self, nums: List[int], k: int) -> int:
        '''
        可以變換的概念有點類似將目前的數組分段，等於每段中都要找到該段最大的元素 然後去計算 mx * len - sum(subarray)的總和
        這可以用前綴和來解
        '''
        n = len(nums)
        f = [[inf] * (k + 1) for _ in range(n + 1)]
        f[0][0] = 0
        mx, tot = -inf, 0
        for i in range(n):
            if nums[i] > mx: mx = nums[i]
            tot += nums[i]
            f[i+1][0] = mx * (i + 1) - tot
        for i in range(n):
            groups = i + 1
            if k < groups: groups = k # 前面可以組成幾組
            for j in range(0, groups):
                mx, tot = -inf, 0
                for t in range(i, -1, -1):
                    if t + 1 < j: break # 預先停止 如果使用的組數少於目前預測前面可以組成的數量就直接break
                    if nums[t] > mx: mx = nums[t]
                    tot += nums[t]
                    if f[t][j] + mx * (i - t + 1) - tot < f[i+1][j+1]:
                        f[i+1][j+1] = f[t][j] + mx * (i - t + 1) - tot
        return f[n][k]

        '''
        遞歸
        @cache 
        def dfs(i:int, j:int):
            if i < 0:
                return 0
            if j == 0:
                return max(nums[:i+1]) * (i + 1) - sum(nums[:i+1])
            ans, mx = inf, -inf
            cur_sum = 0
            for t in range(i, -1, -1):
                if nums[t] > mx: mx = nums[t]
                cur_sum += nums[t]
                nxt = dfs(t - 1, j - 1)
                if nxt + mx * (i - t + 1) - cur_sum < ans: ans = nxt + mx * (i - t + 1) - cur_sum
            return ans
        return dfs(n-1, k)
        '''
        

        '''
        Presum + 遞歸
        presum = [0] + list(accumulate(nums))
        # print(presum)
        @cache 
        def dfs(i:int, j:int):
            if i == n:
                return 0
            if j == k:
                return max(nums[i:n]) * (n - i) - (presum[n] - presum[i])
            ans, mx = inf, -inf
            for t in range(i, n):
                if nums[t] > mx: mx = nums[t]
                nxt = dfs(t + 1, j + 1)
                if nxt + mx * (t - i + 1) - (presum[t+1] - presum[i]) < ans: ans = nxt + mx * (t - i + 1) - (presum[t+1] - presum[i])
            return ans
        return dfs(0, 0)
        '''
